certbot-ci/certbot_integration_tests/utils/misc.py

```diff
@@ -138,26 +139,25 @@ def generate_test_file_hooks(config_dir, hook_probe):
         except OSError as error:
             if error.errno != errno.EEXIST:
                 raise
-        hook_path = os.path.join(hook_dir, 'hook.{0}'.format(extension))
-        if extension == 'sh':
-            data = '''\
-#!/bin/bash -xe
-if [ "$0" = "{0}" ]; then
-    if [ -z "$RENEWED_DOMAINS" -o -z "$RENEWED_LINEAGE" ]; then
-        echo "Environment variables not properly set!" >&2
-        exit 1
-    fi
-fi
-echo $(basename $(dirname "$0")) >> "{1}"\
-'''.format(renewal_deploy_hook_path, hook_probe)
+
```

Comment:
And... we are good for a new journey in the wonderful "Windows, why ?" train!

So on Linux you have this really nice feature of shebang, that is read by the program loader and allows basically to treat any script as an executable as soon as you have the corresponding interpreter available on your system. 

There is no such things on Windows. The only files accepted by the program loader are pre-built binaries compiled against the win32 API, or Batch/Powershell scripts (well, you have dozens of supported extensions in fact but all lead to that two cases at the end).

This is the primary reason why people invented executable python wrappers like https://bitbucket.org/vinay.sajip/simple_launcher/src/default/, that are typically used by setuptools or distutil on windows. It consists in a fair pre-built binary executable as compiled against win32 API, _appended_ with the python script you want to execute in the very same file. The built logic self inspect its content, recognize the part corresponding to the python script, find the shebang in it, then launch the corresponding executable python launcher installed in the system to feed to it the embedded python script as the command argument. Magic.

What you are refering here has nothing to do with the process loader. It is a convenient shortcut to select the program you want to use and pass the given file to it when you double click on it. What will be done with the given file depends on the program you run: most of the time, it is about opening an editor (text, code, xls/doc) and start immediately the edition on that file. You have something similar with Gnome and other Linux window manager. But it has nothing to do with what will happen when you call `subprocess.Popen` that is mapped to the process launcher of the underlying system, and so have the limitation on Windows that I just described.

Reply:
Now the context is settled.

My implementation here is to leverage the capability of executing Batch scripts out-of-the-box on Windows, and so craft a wrapper of python interpreter + hook to launch into a Batch script. This way `subprocess.Popen` and Windows behind are happy.

As this approach is also possible on Linux, I did the same for consistency and readability purpose.

Reply:
Interesting. I was basing this on having read https://docs.python.org/3/faq/windows.html#how-do-i-make-python-scripts-executable the other day.

Testing this in my Windows VM, just executing the Python script directly with the subprocess module works there if `shell=True` like it is [when we execute hooks](https://github.com/certbot/certbot/blob/14e10f40e5cabc3377e49385871b347f5723c8b2/certbot/hooks.py#L247). Should we not expect this behavior to be reliable on other Window systems?

Reply:
I was not aware of this behavior. Last time I tried (at the beginning of the year), the only way to execute a Python script was to pass it to the python interpreter from the command line, or to use a wrapper.

Reply:
It seems to be quite recent. Almost all documentation I could find on internet is not aware of this, and say to call explicitly the interpreter instead.

Reply:
OK.

While removing the batch and shell scripts makes the code a lot simpler, I'm fine keeping things as is if you are so tests should work regardless of the Python installer you use.

Reply:
By the way, why are we invoking the hooks using `shell=True`? This is strongly discouraged by the official documentation, and  in fact, we are in the worst situation that is described to discouraged this usage: a call to an arbitrary command passed to a python program (here the hook command line through the dedicated flags).

Reply:
Beside this, yes I think we should keep the wrapper, at least for now, until we now exactly since when the installer do this hack to make python scripts directly executable, checking that it would be deployed to an existing python installation upon update, and which precise minimal version of Python certbot will support on windows.

Reply:
> By the way, why are we invoking the hooks using shell=True? 

There's a brief discussion of why including `shell=True` is nice for users at https://github.com/certbot/certbot/issues/2676#issuecomment-198087749 but it doesn't get into the potential security concerns.

I don't think it's a problem though. The danger of the `shell` option is [shell injection](https://en.wikipedia.org/wiki/Code_injection#Shell_injection) as described at https://docs.python.org/3.7/library/subprocess.html#security-considerations. I think this is a real problem if you're trying to use user input in a limited way, such as passing user defined arguments to another program.

Certbot, however, is literally letting users define arbitrary commands and we'll run it for them. They don't need to do any fancy shell injection. If they're able to control the string of the hook, they can put any malicious code there directly.

> yes I think we should keep the wrapper, at least for now

Sounds good.

Reply:
Thanks, it makes sense!